### ISSUE TITLE
LWJGL sometimes loads wrong libGL.so.1 on Linux unless -Djava.library.path set to anything

### ISSUE THREAD

**Issue by cuchaz** [LWJGL/lwjgl3#291]
This one took me a while to figure out, but here we go:

Here's a minimal test Java program:
```java
package anything;

import org.lwjgl.glfw.Callbacks;
import org.lwjgl.glfw.GLFW;
import org.lwjgl.opengl.GL;
import org.lwjgl.opengl.GL11;
import org.lwjgl.system.Configuration;
import org.lwjgl.system.MemoryUtil;

public class TestMain {
	
	public static void main(String[] args) {
		
		Configuration.DEBUG.set(true);
		
		GLFW.glfwInit();
		long hwnd = GLFW.glfwCreateWindow(600, 338, "LWJGL3 Test", MemoryUtil.NULL, MemoryUtil.NULL);
		
		GLFW.glfwMakeContextCurrent(hwnd);
		GL.createCapabilities();
		System.out.println("OpenGL Initialized: " + GL11.glGetString(GL11.GL_VERSION));
		
		Callbacks.glfwFreeCallbacks(hwnd);
		GLFW.glfwDestroyWindow(hwnd);
		GLFW.glfwTerminate();
	}
}
```

With no `-Djava.library.path` set, we get this output:
```
[LWJGL] Version: 3.1.1 build 16
[LWJGL] 	 OS: Linux v4.4.0-57-generic
[LWJGL] 	JRE: 1.8.0_121 amd64
[LWJGL] 	JVM: OpenJDK 64-Bit Server VM v25.121-b13 by Oracle Corporation
[LWJGL] Loading library (system): lwjgl
[LWJGL] 	Found at: /tmp/lwjgljeff/3.1.1-build-16/liblwjgl.so
[LWJGL] 	Loaded from org.lwjgl.librarypath: /tmp/lwjgljeff/3.1.1-build-16/liblwjgl.so
[LWJGL] Loading library: glfw
[LWJGL] 	Found at: /tmp/lwjgljeff/3.1.1-build-16/libglfw.so
[LWJGL] MemoryUtil accessor: MemoryAccessorUnsafe
[LWJGL] 	Loaded from org.lwjgl.librarypath: /tmp/lwjgljeff/3.1.1-build-16/libglfw.so
[LWJGL] Loading library (system): lwjgl_opengl
[LWJGL] 	Found at: /tmp/lwjgljeff/3.1.1-build-16/liblwjgl_opengl.so
[LWJGL] 	Loaded from org.lwjgl.librarypath: /tmp/lwjgljeff/3.1.1-build-16/liblwjgl_opengl.so
[LWJGL] Loading library: libGL.so.1
[LWJGL] 	libGL.so.1 not found in org.lwjgl.librarypath=/tmp/lwjgljeff/3.1.1-build-16
[LWJGL] 	Loaded from java.library.path: /usr/lib/libGL.so.1
Exception in thread "main" java.lang.IllegalStateException: There is no OpenGL context current in the current thread.
	at org.lwjgl.opengl.GL.createCapabilities(GL.java:350)
	at org.lwjgl.opengl.GL.createCapabilities(GL.java:302)
	at anything.TestMain.main(TestMain.java:20)
```

When we set the `java.library.path` to anything at all, *even a path that doesn't exist* (e.g., `-Djava.library.path=nope`), we get the desired result:
```
[LWJGL] Version: 3.1.1 build 16
[LWJGL] 	 OS: Linux v4.4.0-57-generic
[LWJGL] 	JRE: 1.8.0_121 amd64
[LWJGL] 	JVM: OpenJDK 64-Bit Server VM v25.121-b13 by Oracle Corporation
[LWJGL] Loading library (system): lwjgl
[LWJGL] 	Found at: /tmp/lwjgljeff/3.1.1-build-16/liblwjgl.so
[LWJGL] 	Loaded from org.lwjgl.librarypath: /tmp/lwjgljeff/3.1.1-build-16/liblwjgl.so
[LWJGL] Loading library: glfw
[LWJGL] 	Found at: /tmp/lwjgljeff/3.1.1-build-16/libglfw.so
[LWJGL] MemoryUtil accessor: MemoryAccessorUnsafe
[LWJGL] 	Loaded from org.lwjgl.librarypath: /tmp/lwjgljeff/3.1.1-build-16/libglfw.so
[LWJGL] Loading library (system): lwjgl_opengl
[LWJGL] 	Found at: /tmp/lwjgljeff/3.1.1-build-16/liblwjgl_opengl.so
[LWJGL] 	Loaded from org.lwjgl.librarypath: /tmp/lwjgljeff/3.1.1-build-16/liblwjgl_opengl.so
[LWJGL] Loading library: libGL.so.1
[LWJGL] 	libGL.so.1 not found in org.lwjgl.librarypath=/tmp/lwjgljeff/3.1.1-build-16
[LWJGL] 	libGL.so.1 not found in java.library.path=nope
[LWJGL] 	Loaded from system paths
OpenGL Initialized: 4.5.0 NVIDIA 378.13
```

Looks like my system has quite a few `libGL.so.1` files laying around:
```
$ find . -name libGL.so.1
./usr/lib/libGL.so.1
./usr/lib/x86_64-linux-gnu/mesa/libGL.so.1
./usr/lib/i386-linux-gnu/libGL.so.1
./usr/lib/i386-linux-gnu/mesa/libGL.so.1
./usr/lib/nvidia-378/libGL.so.1
./usr/lib32/libGL.so.1
./usr/lib32/nvidia-378/libGL.so.1
```

`/usr/lib/` is apparently the wrong choice for my system. Looks like it's a symlink to an old AMD driver: `/usr/lib/fglrx/fglrx-libGL.so.1.2` that's not used anymore. I don't actually know which one is the right choice, but I'd guess it's the one from the NVidia driver: `/usr/lib/nvidia-378/`.

We should probably just let the OS pick the library instead of trying to guess the correct one by searching `java.library.path` manually in LWJGL?

If I don't set `java.library.path` it apparently is set to this by default:
```
/usr/java/packages/lib/amd64:/usr/lib/x86_64-linux-gnu/jni:/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu:/usr/lib/jni:/lib:/usr/lib
```

It looks like whatever magic `DynamicLinkLoader.dlopen()` does is the right thing in this case. I'm guessing it queries the OS using the short library name?

Also, this was very difficult to debug. Any chance there's a way to detect if the wrong OpenGL library was loaded and give a more helpful error message? Maybe just show the actual OpenGL library path (and resolve any symlinks) in the error message? Seeing `fglrx` in the error message would have tipped me off right away that it was a library loading problem.

**Comment by Spasi** [comment]
Thanks!

**Comment by raould** [comment]
here i am in 2018 wishing for better docs around this somewhere, anywhere i could find on the internet, since i am hitting a runtime exception java.lang.UnsatisfiedLinkError: Failed to locate library: libGL.so.1 and have yet to suss out how to resolve it :-)

**Comment by Spasi** [comment]
@raould Could you share more information about your setup? LWJGL version, OS, GPU, driver version, does the native glxgears application work? Also, what does LWJGL output if you run with `-Dorg.lwjgl.util.Debug=true -Dorg.lwjgl.util.DebugLoader=true`?

**Comment by raould** [comment]
thanks...

java.vm.version 25.181-b13 OpenJDK 64-bit Server VM
lwjgl 3.1.2
ubuntu 16.04 lts 64 bit
geforce gtx 480
os claims "using nvidia binary driver 384.130 from nvidia-384"

 /usr/lib32/nvidia-384/libGL.so.1                                                                                                                                                    
 /usr/lib/i386-linux-gnu/mesa/libGL.so.1                                                                                                                                             
 /usr/lib/nvidia-384/libGL.so.1                                                                                                                                                      
 /usr/lib/x86_64-linux-gnu/mesa/libGL.so.1   

I have a version that runs, via some magic (using mill build tool for scala) vs. not working when i try to do it with raw java command with "-D" and "-cp" flags. Somehow the broken one is loading a different LWJGL, 3.2.0 whereas the working one is using 3.1.2. I have no idea where the 3.2.0 is coming from, the only jars on the classpath afaict are the 3.1.2 ones. I will keep trying to figure it out.

good
```
INFO: Running on jMonkeyEngine 3.2-stable
  * Branch: HEAD
  * Git Hash: f85624a
  * Build Date: 2018-01-21
 [LWJGL] Version: 3.1.2 build 29
 [LWJGL]      OS: Linux v4.15.0-36-generic
 [LWJGL]     JRE: 1.8.0_181 amd64
 [LWJGL]     JVM: OpenJDK 64-Bit Server VM v25.181-b13 by Oracle Corporation
 [LWJGL] Loading library (system): lwjgl
 [LWJGL]     Using SharedLibraryLoader...
 [LWJGL]     Found at: /tmp/lwjglsuperman/3.1.2-build-29/liblwjgl.so
 [LWJGL]     Loaded from org.lwjgl.librarypath: /tmp/lwjglsuperman/3.1.2-build-29/liblwjgl.so
 [LWJGL] MemoryUtil accessor: MemoryAccessorUnsafe
 [LWJGL] Loading library: jemalloc
 [LWJGL]     Using SharedLibraryLoader...
 [LWJGL]     Found at: /tmp/lwjglsuperman/3.1.2-build-29/libjemalloc.so
 [LWJGL]     Loaded from org.lwjgl.librarypath: /tmp/lwjglsuperman/3.1.2-build-29/libjemalloc.so
 [LWJGL] MemoryUtil allocator: JEmallocAllocator
 [LWJGL] Loading library: glfw
 [LWJGL]     Using SharedLibraryLoader...
 [LWJGL]     Found at: /tmp/lwjglsuperman/3.1.2-build-29/libglfw.so
 [LWJGL]     Loaded from org.lwjgl.librarypath: /tmp/lwjglsuperman/3.1.2-build-29/libglfw.so
 Oct 09, 2018 8:54:42 PM com.jme3.system.lwjgl.LwjglContext printContextInitInfo
 INFO: LWJGL 3.1.2 build 29 context running on thread main
  * Graphics Adapter: GLFW 3.3.0 X11 GLX EGL clock_gettime /dev/js shared
 [LWJGL] Loading library (system): lwjgl_opengl
 [LWJGL]     Using SharedLibraryLoader...
 [LWJGL]     Found at: /tmp/lwjglsuperman/3.1.2-build-29/liblwjgl_opengl.so
 [LWJGL]     Loaded from org.lwjgl.librarypath: /tmp/lwjglsuperman/3.1.2-build-29/liblwjgl_opengl.so
 [LWJGL] Loading library: libGL.so.1
 [LWJGL]     libGL.so.1 not found in org.lwjgl.librarypath=/tmp/lwjglsuperman/3.1.2-build-29
 [LWJGL]     Loaded from system paths
 Oct 09, 2018 8:54:42 PM com.jme3.renderer.opengl.GLRenderer loadCapabilitiesCommon
 INFO: OpenGL Renderer Information
  * Vendor: NVIDIA Corporation
  * Renderer: GeForce GTX 480/PCIe/SSE2
  * OpenGL Version: 4.5.0 NVIDIA 384.130
  * GLSL Version: 4.50 NVIDIA
  * Profile: Compatibility
```

bad
```
INFO: Running on jMonkeyEngine 3.2-stable
  * Branch: HEAD
  * Git Hash: f85624a
  * Build Date: 2018-01-21
 [LWJGL] Version: 3.2.0 build 12
 [LWJGL]      OS: Linux v4.15.0-36-generic
 [LWJGL]     JRE: 1.8.0_181 amd64
 [LWJGL]     JVM: OpenJDK 64-Bit Server VM v25.181-b13 by Oracle Corporation
 [LWJGL] Loading library (system): lwjgl
 [LWJGL]     Using SharedLibraryLoader...
 [LWJGL]     Found at: /tmp/lwjglsuperman/3.2.0-build-12/liblwjgl.so
 [LWJGL]     Loaded from org.lwjgl.librarypath: /tmp/lwjglsuperman/3.2.0-build-12/liblwjgl.so
 [LWJGL] MemoryUtil accessor: MemoryAccessorUnsafe
 [LWJGL] Loading library: jemalloc
 [LWJGL]     Using SharedLibraryLoader...
 [LWJGL]     Found at: /tmp/lwjglsuperman/3.2.0-build-12/libjemalloc.so
 [LWJGL]     Loaded from org.lwjgl.librarypath: /tmp/lwjglsuperman/3.2.0-build-12/libjemalloc.so
 [LWJGL] MemoryUtil allocator: JEmallocAllocator
 [LWJGL] Loading library: glfw
 [LWJGL]     Using SharedLibraryLoader...
 [LWJGL]     Found at: /tmp/lwjglsuperman/3.2.0-build-12/libglfw.so
 [LWJGL]     Loaded from org.lwjgl.librarypath: /tmp/lwjglsuperman/3.2.0-build-12/libglfw.so
 Oct 09, 2018 9:02:59 PM com.jme3.system.lwjgl.LwjglContext printContextInitInfo
 INFO: LWJGL 3.2.0 build 12 context running on thread main
  * Graphics Adapter: GLFW 3.3.0 X11 GLX EGL clock_gettime /dev/js shared
 [LWJGL] Loading library (system): lwjgl_opengl
 [LWJGL]     Using SharedLibraryLoader...
 [LWJGL]     Found at: /tmp/lwjglsuperman/3.2.0-build-12/liblwjgl_opengl.so
 [LWJGL]     Loaded from org.lwjgl.librarypath: /tmp/lwjglsuperman/3.2.0-build-12/liblwjgl_opengl.so
 [LWJGL] Loading library: libGL.so.1
 [LWJGL]     libGL.so.1 not found in org.lwjgl.librarypath=/tmp/lwjglsuperman/3.2.0-build-12
 [LWJGL]     libGL.so.1 not found in system paths
 [LWJGL]     libGL.so.1 not found in java.library.path=/usr/java/packages/lib/amd64:/usr/lib/x86_64-linux-gnu/jni:/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu:/usr/lib/jni:/lib:/u\
 sr/lib
 [LWJGL] Failed to load a library. Possible solutions:
     a) Install the library or the driver that provides the library.
     b) Ensure that the library is accessible from the system library paths.
 [LWJGL] Loading library: libGL.so
 [LWJGL]     libGL.so not found in org.lwjgl.librarypath=/tmp/lwjglsuperman/3.2.0-build-12
 [LWJGL]     libGL.so not found in system paths
 [LWJGL]     libGL.so not found in java.library.path=/usr/java/packages/lib/amd64:/usr/lib/x86_64-linux-gnu/jni:/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu:/usr/lib/jni:/lib:/usr\
 /lib
 [LWJGL] Failed to load a library. Possible solutions:
     a) Install the library or the driver that provides the library.
     b) Ensure that the library is accessible from the system library paths.
 Oct 09, 2018 9:02:59 PM com.jme3.app.LegacyApplication handleError
 SEVERE: Uncaught exception thrown in Thread[main,5,main]
 java.lang.UnsatisfiedLinkError: Failed to locate library: libGL.so.1
```

**Comment by raould** [comment]
p.s. glxgears works fine. :-)

**Comment by raould** [comment]
p.s. glxgears works fine. :-)

**Comment by raould** [comment]
ok fixed, i somehow managed to exorcise all the 3.2.0 (lwjgl and opengl) jars. phew. weird that 3.2.0 is worse than 3.1.2?

**Comment by Spasi** [comment]
@raould It is weird indeed (specifically this line: `libGL.so.1 not found in system paths`). It would be helpful if you could test a clean project (without jMonkeyEngine, mixed LWJGL versions, etc). Try cloning [lwjgl3-demos](https://github.com/LWJGL/lwjgl3-demos) and running one of the OpenGL samples.
